File: backend/app/clients/govinfo.py
import asyncio
import logging
from typing import Optional
from datetime import datetime, timedelta, timezone
from email.utils import parsedate_to_datetime
import httpx

try:
    import h2  # noqa: F401
    _HTTP2_AVAILABLE = True
except ImportError:
    _HTTP2_AVAILABLE = False

from .base import BaseAPIClient
from .html_utils import html_to_text
from .pdf_utils import extract_pdf_images_sync, extract_pdf_text_sync
from ..config import get_settings
from ..models.schemas import SourceItem

logger = logging.getLogger(__name__)


class GovInfoClient(BaseAPIClient):
    _shared_content_clients: dict[float, httpx.AsyncClient] = {}

    def __init__(self):
        settings = get_settings()
        super().__init__(base_url=settings.govinfo_base_url)
        self.api_key = settings.gov_api_key
        self._content_client = self._get_content_client(self.timeout)

    @classmethod
    def _get_content_client(cls, timeout: float) -> httpx.AsyncClient:
        client = cls._shared_content_clients.get(timeout)
        if client:
            return client
        client = httpx.AsyncClient(
            timeout=timeout,
            follow_redirects=True,
            http2=_HTTP2_AVAILABLE,
            limits=httpx.Limits(max_connections=10, max_keepalive_connections=5),
        )
        cls._shared_content_clients[timeout] = client
        return client

    @classmethod
    async def close_shared_clients(cls) -> None:
        for client in cls._shared_content_clients.values():
            await client.aclose()
        cls._shared_content_clients = {}

    def _parse_retry_after(self, value: Optional[str]) -> Optional[float]:
        if not value:
            return None

        try:
            return float(value)
        except ValueError:
            try:
                parsed = parsedate_to_datetime(value)
                if parsed.tzinfo is None:
                    parsed = parsed.replace(tzinfo=timezone.utc)
                now = datetime.now(timezone.utc)
                return max(0.0, (parsed - now).total_seconds())
            except Exception:
                return None

    async def _fetch_url_with_retry(
        self,
        client: httpx.AsyncClient,
        url: str,
        params: Optional[dict],
    ) -> Optional[httpx.Response]:
        backoff = self.settings.initial_backoff
        max_attempts = self.settings.max_retries + 1

        for attempt in range(max_attempts):
            try:
                response = await client.get(url, params=params)
            except httpx.TimeoutException:
                if attempt < max_attempts - 1:
                    await asyncio.sleep(backoff)
                    backoff = min(backoff * 2, 30.0)
                    continue
                return None
            except httpx.RequestError:
                if attempt < max_attempts - 1:
                    await asyncio.sleep(backoff)
                    backoff = min(backoff * 2, 30.0)
                    continue
                return None

            if response.status_code == 429:
                retry_after = self._parse_retry_after(
                    response.headers.get("Retry-After")
                )
                if attempt < max_attempts - 1:
                    await asyncio.sleep(retry_after or backoff)
                    backoff = min(backoff * 2, 30.0)
                    continue
                return response

            if response.status_code == 408 or response.status_code >= 500:
                if attempt < max_attempts - 1:
                    await asyncio.sleep(backoff)
                    backoff = min(backoff * 2, 30.0)
                    continue
                return response

            return response

        return None

    def _looks_like_text(self, content: bytes) -> bool:
        if not content:
            return True
        sample = content[:1024]
        if b"\x00" in sample:
            return False
        printable = sum(
            1 for b in sample
            if 32 <= b <= 126 or b in (9, 10, 13)
        )
        return printable / len(sample) > 0.85

    def _is_supported_text_type(self, content_type: str, content: bytes) -> bool:
        if not content_type:
            return self._looks_like_text(content)

        content_type = content_type.lower()
        if content_type.startswith("text/"):
            return True
        if "html" in content_type or "xml" in content_type or "json" in content_type:
            return True
        if "octet-stream" in content_type or "binary" in content_type:
            return self._looks_like_text(content)
        return False

    def _add_api_key(self, params: Optional[dict] = None) -> dict:
        params = params or {}
        params["api_key"] = self.api_key
        return params

    def _normalize_search_result(self, result: dict) -> SourceItem:
        package_id = result.get("packageId", "")
        title = result.get("title", "Untitled")

        url = f"https://www.govinfo.gov/app/details/{package_id}"
        if result.get("granuleId"):
            url = f"https://www.govinfo.gov/app/details/{package_id}/{result['granuleId']}"

        date = result.get("lastModified") or result.get("dateIssued")

        agency = None
        if result.get("governmentAuthor"):
            authors = result.get("governmentAuthor", [])
            if isinstance(authors, list) and authors:
                agency = authors[0]
            elif isinstance(authors, str):
                agency = authors

        return SourceItem(
            source_type="govinfo_result",
            id=package_id,
            title=title,
            agency=agency,
            date=date,
            url=url,
            excerpt=result.get("abstract") or result.get("description"),
        )

    def _normalize_package(self, package: dict) -> SourceItem:
        package_id = package.get("packageId", "")
        title = package.get("title", "Untitled Package")

        url = f"https://www.govinfo.gov/app/details/{package_id}"

        return SourceItem(
            source_type="govinfo_package",
            id=package_id,
            title=title,
            agency=package.get("publisher"),
            date=package.get("lastModified") or package.get("dateIssued"),
            url=url,
            excerpt=package.get("abstract") or package.get("description"),
        )

    async def search(
        self,
        query: str,
        page_size: int = 10,
        offset_mark: str = "*",
        sorts: Optional[list[dict]] = None,
    ) -> tuple[dict, list[SourceItem]]:
        if sorts is None:
            sorts = [{"field": "lastModified", "sortOrder": "DESC"}]

        url = f"{self.base_url}/search"
        params = self._add_api_key()

        body = {
            "query": query,
            "pageSize": str(page_size),
            "offsetMark": offset_mark,
            "sorts": sorts,
        }

        logger.info(f"Searching GovInfo: {query}")

        data = await self._request_with_retry(
            method="POST",
            url=url,
            params=params,
            json=body,
            use_cache=False,
        )

        results = data.get("results", [])
        sources = [self._normalize_search_result(r) for r in results]

        return data, sources

    async def get_package_summary(
        self, package_id: str
    ) -> tuple[dict, SourceItem]:
        url = f"{self.base_url}/packages/{package_id}/summary"
        params = self._add_api_key()

        logger.info(f"Fetching GovInfo package: {package_id}")

        data = await self._request_with_retry(
            method="GET",
            url=url,
            params=params,
        )

        source = self._normalize_package(data)

        return data, source

    async def get_package_content(
        self, package_id: str, max_length: int = 15000
    ) -> tuple[str, SourceItem, list[dict], int, str, Optional[str]]:
        summary, source = await self.get_package_summary(package_id)

        url = f"{self.base_url}/packages/{package_id}/htm"
        params = self._add_api_key()

        logger.info(f"Fetching GovInfo package content: {package_id}")
        images: list[dict] = []
        images_skipped = 0
        content_format = "unknown"
        pdf_url = f"{self.base_url}/packages/{package_id}/pdf"

        try:
            text_result = ""

            client = self._content_client
            for fmt in ("htm", "xml", "txt"):
                if fmt == "htm":
                    target_url = url
                else:
                    target_url = f"{self.base_url}/packages/{package_id}/{fmt}"

                response = await self._fetch_url_with_retry(
                    client,
                    target_url,
                    params,
                )
                if not response or response.status_code != 200:
                    continue

                content_type = response.headers.get("content-type", "")
                if not self._is_supported_text_type(content_type, response.content):
                    continue

                text = html_to_text(response.text, max_length)
                if text:
                    text_result = text
                    content_format = fmt
                    break

            if text_result:
                should_extract_images = self.settings.pdf_extract_images
                if should_extract_images:
                    pdf_response = await self._fetch_url_with_retry(
                        client,
                        pdf_url,
                        params,
                    )
                    if pdf_response and pdf_response.status_code == 200:
                        extracted_images, skipped = await asyncio.to_thread(
                            extract_pdf_images_sync,
                            pdf_response.content,
                        )
                        logger.info(
                            "GovInfo PDF images extracted for %s: images=%s skipped=%s",
                            package_id,
                            len(extracted_images),
                            skipped,
                        )
                        if extracted_images:
                            images = extracted_images
                            images_skipped = skipped

                return text_result, source, images, images_skipped, content_format, pdf_url

            pdf_response = await self._fetch_url_with_retry(
                client,
                pdf_url,
                params,
            )
            if pdf_response and pdf_response.status_code == 200:
                pdf_text = await asyncio.to_thread(
                    extract_pdf_text_sync,
                    pdf_response.content,
                    max_length,
                )
                should_extract_images = self.settings.pdf_extract_images or not pdf_text
                if should_extract_images:
                    extracted_images, skipped = await asyncio.to_thread(
                        extract_pdf_images_sync,
                        pdf_response.content,
                    )
                    logger.info(
                        "GovInfo PDF extracted for %s: text=%s images=%s skipped=%s",
                        package_id,
                        len(pdf_text) if pdf_text else 0,
                        len(extracted_images),
                        skipped,
                    )
                    if extracted_images:
                        images = extracted_images
                        images_skipped = skipped
                if pdf_text or images:
                    content_format = "pdf"
                    return pdf_text or "", source, images, images_skipped, content_format, pdf_url

        except Exception as e:
            logger.warning(f"Could not fetch content for {package_id}: {e}")

        fallback = summary.get("abstract") or summary.get("description") or "No content available."
        return fallback, source, images, images_skipped, content_format, pdf_url

    async def get_collection(
        self,
        collection_code: str,
        start_datetime: Optional[str] = None,
        page_size: int = 10,
        offset_mark: str = "*",
    ) -> tuple[dict, list[SourceItem]]:
        if start_datetime is None:
            start = datetime.utcnow() - timedelta(days=30)
            start_datetime = start.strftime("%Y-%m-%dT%H:%M:%SZ")

        url = f"{self.base_url}/collections/{collection_code}/{start_datetime}"
        params = self._add_api_key({
            "pageSize": page_size,
            "offsetMark": offset_mark,
        })

        logger.info(f"Fetching GovInfo collection: {collection_code}")

        data = await self._request_with_retry(
            method="GET",
            url=url,
            params=params,
        )

        packages = data.get("packages", [])
        sources = [self._normalize_package(p) for p in packages]

        return data, sources


def build_govinfo_query(
    keywords: str,
    collection: Optional[str] = None,
    days: Optional[int] = None,
) -> str:
    keywords = (keywords or "").strip()
    parts = []
    lowered = keywords.lower()

    if collection and "collection:" not in lowered:
        parts.append(f"collection:{collection}")

    if keywords:
        parts.append(keywords)

    if days and "publishdate:range" not in lowered and "dateissued:range" not in lowered:
        start_date = (datetime.utcnow() - timedelta(days=days)).strftime("%Y-%m-%d")
        parts.append(f"publishdate:range({start_date},)")

    return " AND ".join(parts)

File: backend-ts/src/clients/govinfo.ts
import { BaseAPIClient } from "./base.js";
import { getSettings } from "../config.js";
import type { SourceItem } from "../models/schemas.js";
import { extractPdfText } from "./pdfUtils.js";

export function htmlToText(html: string, maxLength: number = 15000): string {
    let text = html.replace(/<script[^>]*>[\s\S]*?<\/script>/gi, "");
    text = text.replace(/<style[^>]*>[\s\S]*?<\/style>/gi, "");

    text = text.replace(/<\/(p|div|h[1-6]|li|tr|br)[^>]*>/gi, "\n");
    text = text.replace(/<(br|hr)[^>]*\/?>/gi, "\n");

    text = text.replace(/<[^>]+>/g, "");

    text = text.replace(/&nbsp;/g, " ");
    text = text.replace(/&amp;/g, "&");
    text = text.replace(/&lt;/g, "<");
    text = text.replace(/&gt;/g, ">");
    text = text.replace(/&quot;/g, '"');
    text = text.replace(/&#39;/g, "'");

    text = text.replace(/[ \t]+/g, " ");
    text = text.replace(/\n\s*\n/g, "\n\n");
    text = text.trim();

    if (text.length > maxLength) {
        text = text.slice(0, maxLength) + "\n\n[Content truncated due to length...]";
    }

    return text;
}

export class GovInfoClient extends BaseAPIClient {
    private apiKey: string;

    constructor() {
        const settings = getSettings();
        super(settings.govInfoBaseUrl);
        this.apiKey = settings.govApiKey;
    }

    private addApiKey(params: Record<string, string | number> = {}): Record<string, string | number> {
        return { ...params, api_key: this.apiKey };
    }

    private normalizeSearchResult(result: Record<string, unknown>): SourceItem {
        const packageId = result.packageId as string || "";
        const title = result.title as string || "Untitled";

        let url = `https://www.govinfo.gov/app/details/${packageId}`;
        if (result.granuleId) {
            url = `https://www.govinfo.gov/app/details/${packageId}/${result.granuleId}`;
        }

        const date = (result.lastModified as string) || (result.dateIssued as string) || null;

        let agency: string | null = null;
        const authors = result.governmentAuthor;
        if (Array.isArray(authors) && authors.length > 0) {
            agency = authors[0] as string;
        } else if (typeof authors === "string") {
            agency = authors;
        }

        return {
            source_type: "govinfo_result",
            id: packageId,
            title,
            agency,
            date,
            url,
            excerpt: (result.abstract as string) || (result.description as string) || null,
        };
    }

    private normalizePackage(pkg: Record<string, unknown>): SourceItem {
        const packageId = pkg.packageId as string || "";
        const title = pkg.title as string || "Untitled Package";

        const url = `https://www.govinfo.gov/app/details/${packageId}`;

        return {
            source_type: "govinfo_package",
            id: packageId,
            title,
            agency: pkg.publisher as string || null,
            date: (pkg.lastModified as string) || (pkg.dateIssued as string) || null,
            url,
            excerpt: (pkg.abstract as string) || (pkg.description as string) || null,
        };
    }

    async search(options: {
        query: string;
        pageSize?: number;
        offsetMark?: string;
        sorts?: { field: string; sortOrder: string }[];
    }): Promise<{ data: Record<string, unknown>; sources: SourceItem[] }> {
        const {
            query,
            pageSize = 10,
            offsetMark = "*",
            sorts = [{ field: "lastModified", sortOrder: "DESC" }],
        } = options;

        const url = `${this.baseUrl}/search`;
        const params = this.addApiKey();

        const data = await this.requestWithRetry<Record<string, unknown>>({
            method: "POST",
            url,
            params,
            json: {
                query,
                pageSize: String(pageSize),
                offsetMark,
                sorts,
            },
            useCache: false,
        });

        const results = (data.results || []) as Record<string, unknown>[];
        const sources = results.map((r) => this.normalizeSearchResult(r));

        return { data, sources };
    }

    async getPackageSummary(packageId: string): Promise<{ data: Record<string, unknown>; source: SourceItem }> {
        const url = `${this.baseUrl}/packages/${packageId}/summary`;
        const params = this.addApiKey();

        const data = await this.requestWithRetry<Record<string, unknown>>({
            url,
            params,
        });

        const source = this.normalizePackage(data);

        return { data, source };
    }

    async getPackageContent(options: {
        packageId: string;
        maxLength?: number;
    }): Promise<{
        text: string;
        source: SourceItem;
        images: unknown[];
        imagesSkipped: number;
        contentFormat: string;
        pdfUrl: string | null;
    }> {
        const { packageId, maxLength = 15000 } = options;

        const { data: summary, source } = await this.getPackageSummary(packageId);

        const pdfUrl = `${this.baseUrl}/packages/${packageId}/pdf`;

        const formats = ["htm", "xml", "txt"];
        let textResult = "";
        let contentFormat = "unknown";

        for (const fmt of formats) {
            try {
                const contentUrl = new URL(`${this.baseUrl}/packages/${packageId}/${fmt}`);
                contentUrl.searchParams.set("api_key", this.apiKey);
                const response = await fetch(contentUrl.toString(), {
                    headers: { Accept: "text/html,application/xml,text/plain" },
                });

                if (response.ok) {
                    const contentType = response.headers.get("content-type") || "";
                    if (
                        contentType.startsWith("text/") ||
                        contentType.includes("html") ||
                        contentType.includes("xml")
                    ) {
                        const html = await response.text();
                        const text = htmlToText(html, maxLength);
                        if (text) {
                            textResult = text;
                            contentFormat = fmt;
                            break;
                        }
                    }
                }
            } catch (error) {
                console.warn(`Failed to fetch ${fmt} for ${packageId}: ${error}`);
            }
        }

        if (!textResult) {
            try {
                const pdfFetchUrl = new URL(pdfUrl);
                pdfFetchUrl.searchParams.set("api_key", this.apiKey);
                const response = await fetch(pdfFetchUrl.toString());
                if (response.ok) {
                    const pdfBuffer = Buffer.from(await response.arrayBuffer());
                    textResult = await extractPdfText(pdfBuffer, maxLength) || "";
                    contentFormat = "pdf";
                }
            } catch (error) {
                console.warn(`Failed to fetch PDF for ${packageId}: ${error}`);
            }
        }

        if (!textResult) {
            textResult = (summary.abstract as string) || (summary.description as string) || "No content available.";
        }

        return {
            text: textResult,
            source,
            images: [],
            imagesSkipped: 0,
            contentFormat,
            pdfUrl,
        };
    }

    async getCollection(options: {
        collectionCode: string;
        startDatetime?: string;
        pageSize?: number;
        offsetMark?: string;
    }): Promise<{ data: Record<string, unknown>; sources: SourceItem[] }> {
        const {
            collectionCode,
            startDatetime,
            pageSize = 10,
            offsetMark = "*",
        } = options;

        const start = startDatetime || new Date(Date.now() - 30 * 24 * 60 * 60 * 1000).toISOString();

        const url = `${this.baseUrl}/collections/${collectionCode}/${start}`;
        const params = this.addApiKey({
            pageSize,
            offsetMark,
        });

        const data = await this.requestWithRetry<Record<string, unknown>>({
            url,
            params,
        });

        const packages = (data.packages || []) as Record<string, unknown>[];
        const sources = packages.map((p) => this.normalizePackage(p));

        return { data, sources };
    }
}

export function buildGovInfoQuery(options: {
    keywords: string;
    collection?: string;
    days?: number;
}): string {
    const { keywords, collection, days } = options;
    const cleanKeywords = (keywords || "").trim();
    const parts: string[] = [];
    const lowered = cleanKeywords.toLowerCase();

    if (collection && !lowered.includes("collection:")) {
        parts.push(`collection:${collection}`);
    }

    if (cleanKeywords) {
        parts.push(cleanKeywords);
    }

    if (days && !lowered.includes("publishdate:range") && !lowered.includes("dateissued:range")) {
        const startDate = new Date(Date.now() - days * 24 * 60 * 60 * 1000).toISOString().split("T")[0];
        parts.push(`publishdate:range(${startDate},)`);
    }

    return parts.join(" AND ");
}
